File: mvpreaderjava/src/main/java/com/example/mvpreaderjava/di/module/HttpModule.java
package com.example.mvpreaderjava.di.module;

import com.example.mvpreaderjava.Constant;
import com.example.mvpreaderjava.SystemUtil;
import com.example.mvpreaderjava.di.qualifier.JuheUrl;
import com.example.mvpreaderjava.modle.helper.JUHEService;

import java.io.File;
import java.io.IOException;
import java.util.concurrent.TimeUnit;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import okhttp3.Cache;
import okhttp3.CacheControl;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by sw116 on 2018/4/19.
 */
@Module
public class HttpModule {
    @Provides
    @Singleton
    public OkHttpClient.Builder provideOkHttpBuilder() {
        return new OkHttpClient.Builder();
    }

    @Provides
    @Singleton
    public Retrofit.Builder provideRetrofitBuilder() {
        return new Retrofit.Builder();
    }

    @Provides
    @Singleton
    public OkHttpClient providerOkHttpClient(OkHttpClient.Builder builder) {
        File cacheFile = new File(Constant.CACHE_PATH);
        Cache cache = new Cache(cacheFile, Long.valueOf((1024 * 1024 * 50)));
        Interceptor cacheInterceptor = new Interceptor() {
            @Override
            public Response intercept(Chain chain) throws IOException {
                Request request = chain.request();
                if (!SystemUtil.isNetworkConnected()) {
                    request = request.newBuilder()
                            .cacheControl(CacheControl.FORCE_CACHE)
                            .build();
                }
                Response response = chain.proceed(request);
                if (SystemUtil.isNetworkConnected()) {
                    int maxAge = 0;
                    // 有网络时, 不缓存, 最大保存时长为0
                    response.newBuilder()
                            .header("Cache-Control", "public, max-age=" + maxAge)
                            .removeHeader("Pragma")
                            .build();
                } else {
                    // 无网络时，设置超时为4周
                    int maxStale = 60 * 60 * 24 * 28;
                    response.newBuilder()
                            .header("Cache-Control", "public, only-if-cached, max-stale=" + maxStale)
                            .removeHeader("Pragma")
                            .build();
                }
                return response;
            }
        };
        //设置缓存
        builder.addNetworkInterceptor(cacheInterceptor);
        builder.addInterceptor(cacheInterceptor);
        builder.cache(cache);
        //设置超时
        builder.connectTimeout(10, TimeUnit.SECONDS);
        builder.readTimeout(20, TimeUnit.SECONDS);
        builder.writeTimeout(20, TimeUnit.SECONDS);
        //错误重连
        builder.retryOnConnectionFailure(true);
        return builder.build();
    }

    /**
     * 提供聚合数据接口service
     * @param retrofit
     * @return
     */
    @Provides
    @Singleton
    public JUHEService provideJUHEService(@JuheUrl Retrofit retrofit) {
        return retrofit.create(JUHEService.class);
    }

    @Provides
    @Singleton
    @JuheUrl
    public Retrofit provideJUHERetrofit(Retrofit.Builder builder, OkHttpClient client) {
        return provideRetrofit(builder, client, JUHEService.API_BASE_URL);
    }

    public Retrofit provideRetrofit(Retrofit.Builder builder, OkHttpClient client, String url) {
        return builder.baseUrl(url).
                addConverterFactory(GsonConverterFactory.create()).
                addCallAdapterFactory(RxJava2CallAdapterFactory.create()).
                client(client).build();
    }

}

File: app/src/main/java/com/savion/mvpreader/di/module/HttpModule.kt
package com.savion.mvpreader.di.module

import com.savion.mvpreader.Constant
import com.savion.mvpreader.common.SystemUtil
import com.savion.mvpreader.di.qualifier.JuheUrl
import com.savion.mvpreader.model.http.service.JUHEService
import dagger.Module
import dagger.Provides
import okhttp3.*
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.io.File
import java.util.concurrent.TimeUnit
import javax.inject.Singleton

/**
 * Created by savion on 2018/4/10.
 */
@Module
open class HttpModule {
    @Provides
    @Singleton
    fun provideRetrofitBuilder(): Retrofit.Builder {
        return Retrofit.Builder()
    }

    @Provides
    @Singleton
    fun provideHttpBuilder(): OkHttpClient.Builder {
        return OkHttpClient.Builder()
    }

    @Provides
    @Singleton
    fun provideHttpClient(builder: OkHttpClient.Builder): OkHttpClient {
        val cacheFile = File(Constant.CACHE_PATH)
        val cache = Cache(cacheFile, (1024 * 1024 * 50).toLong())
        val cacheInterceptor = Interceptor { chain ->
            var request = chain.request()
            if (!SystemUtil.isNetworkConnected()) {
                request = request.newBuilder()
                        .cacheControl(CacheControl.FORCE_CACHE)
                        .build()
            }
            val response = chain.proceed(request)
            if (SystemUtil.isNetworkConnected()) {
                val maxAge = 0
                // 有网络时, 不缓存, 最大保存时长为0
                response.newBuilder()
                        .header("Cache-Control", "public, max-age=" + maxAge)
                        .removeHeader("Pragma")
                        .build()
            } else {
                // 无网络时，设置超时为4周
                val maxStale = 60 * 60 * 24 * 28
                response.newBuilder()
                        .header("Cache-Control", "public, only-if-cached, max-stale=" + maxStale)
                        .removeHeader("Pragma")
                        .build()
            }
            response
        }
//        Interceptor apikey = new Interceptor() {
//            @Override
//            public Response intercept(Chain chain) throws IOException {
//                Request request = chain.request();
//                request = request.newBuilder()
//                        .addHeader("apikey",Constants.KEY_API)
//                        .build();
//                return chain.proceed(request);
//            }
//        }
//        设置统一的请求头部参数
//        builder.addInterceptor(apikey);
        //设置缓存
        builder.addNetworkInterceptor(cacheInterceptor)
        builder.addInterceptor(cacheInterceptor)
        builder.cache(cache)
        //设置超时
        builder.connectTimeout(10, TimeUnit.SECONDS)
        builder.readTimeout(20, TimeUnit.SECONDS)
        builder.writeTimeout(20, TimeUnit.SECONDS)
        //错误重连
        builder.retryOnConnectionFailure(true)
        return builder.build()
    }

    @Provides
    @Singleton
    @JuheUrl
    fun provideJUHEApi(retrofitBuilder: Retrofit.Builder, httpClient: OkHttpClient): Retrofit {
        return provideRetrofit(retrofitBuilder, httpClient, JUHEService.API_BASE_URL)
    }

    @Provides
    @Singleton
    fun provideJUHERetrofit(@JuheUrl retrofit: Retrofit): JUHEService {
        return retrofit.create(JUHEService::class.java)
    }

    fun provideRetrofit(retrofitBuilder: Retrofit.Builder, httpClient: OkHttpClient, url: String): Retrofit {
        return retrofitBuilder
                .baseUrl(url)
                .client(httpClient)
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addConverterFactory(GsonConverterFactory.create())
                .build()
    }

}